pkg/provider/onepasswordsdk/client.go

```diff
@@ -0,0 +1,349 @@
+		FieldType: onepassword.ItemFieldTypeConcealed,
+	}
+
+	return field
+}
+
+func (p *Provider) PushSecret(ctx context.Context, secret *corev1.Secret, ref esv1.PushSecretData) error {
+	val, ok := secret.Data[ref.GetSecretKey()]
+	if !ok {
+		return fmt.Errorf("secret %s/%s does not contain a key", secret.Namespace, secret.Name)
+	}
+
+	title := ref.GetRemoteKey()
+	providerItem, err := p.findItem(ctx, title)
+	if errors.Is(err, ErrKeyNotFound) {
+		if err = p.createItem(ctx, val, ref); err != nil {
+			return fmt.Errorf("failed to create item: %w", err)
+		}
+
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to find item: %w", err)
+	}
+
+	label := ref.GetProperty()
+	if label == "" {
+		label = "password"
+	}
+
+	mdata, err := metadata.ParseMetadataParameters[PushSecretMetadataSpec](ref.GetMetadata())
+	if err != nil {
+		return fmt.Errorf("failed to parse push secret metadata: %w", err)
+	}
+	if mdata != nil && mdata.Spec.Tags != nil {
+		providerItem.Tags = mdata.Spec.Tags
+	}
+
+	providerItem.Fields, err = updateFieldValue(providerItem.Fields, label, string(val))
+	if err != nil {
+		return fmt.Errorf("failed to update field with value %s: %w", string(val), err)
+	}
+
+	if _, err = p.client.Items().Put(ctx, providerItem); err != nil {
+		return fmt.Errorf("failed to update item: %w", err)
+	}
+
+	return nil
+}
+
+func (p *Provider) GetVault(ctx context.Context, name string) (string, error) {
+	vaults, err := p.client.VaultsAPI.ListAll(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to list vaults: %w", err)
+	}
+
+	for {
+		v, err := vaults.Next()
+		if err != nil {
+			// the only time the iterator returns an error is when it's done.
+			break
+		}
+
+		if v.Title == name {
+			// cache the ID so we don't have to repeat this lookup.
+			p.vaultID = v.ID
+			return v.ID, nil
+		}
+	}
+
+	return "", fmt.Errorf("vault %s not found", name)
+}
+
+func (p *Provider) findItem(ctx context.Context, name string) (onepassword.Item, error) {
+	if strfmt.IsUUID(name) {
+		return p.client.Items().Get(ctx, p.vaultID, name)
+	}
+
+	items, err := p.client.Items().ListAll(ctx, p.vaultID)
+	if err != nil {
+		return onepassword.Item{}, fmt.Errorf("failed to list items: %w", err)
+	}
+
+	// We don't stop
+	var itemUUID string
+	for {
+		v, err := items.Next()
+		// the only time the iterator returns an error is when it's done.
+		if err != nil {
+			break
+		}
+
+		if v.Title == name {
+			if itemUUID != "" {
+				return onepassword.Item{}, fmt.Errorf("found multiple items with name %s", name)
+			}
+			itemUUID = v.ID
+		}
+	}
+
+	if itemUUID == "" {
+		return onepassword.Item{}, ErrKeyNotFound
+	}
+
+	return p.client.Items().Get(ctx, p.vaultID, itemUUID)
+}
+
+// SecretExists Not Implemented.
+func (p *Provider) SecretExists(ctx context.Context, ref esv1.PushSecretRemoteRef) (bool, error) {
+	_, err := p.GetSecret(ctx, esv1.ExternalSecretDataRemoteRef{
+		Key: ref.GetRemoteKey(),
+	})
+
+	return err == nil, nil
+}
```

Comment:
wouldn't this fail to set a given remoteRef.Property that isn't present when the ifNotExists updatePolicy is set?

as in, an 1password item with no `foobar` label, but a PushSecret trying to push that `foobar` label exactly once if that makes sense.

Reply:
I'm not sure I understand, sorry. This isn't checking for a label. It's checking for an entire secret to exist.

RemoteKey is the title of an entire item, while the property is the label. So this basically checks if the entire secret exists or not.

Reply:
Yeah, but that’s what I mean.

imagine to Pushes each with one property. The first one succeeds, the second one fails.

Reply:
You mean each with a different property, right?

Reply:
Yah, the second one would fail, because the remoteKey already exists, but we don't check for property. I think that's fine. There shouldn't be two push secrets with the same title but with different properties. And if you Edit a push secret, that doesn't go into the same flow does it? Is that the scenario you think would fail? ( it probably would yeah, but it's so weird... :D )

Reply:
It’s more because we don’t allow pushing the whole secret. This means we would need to add multiple entries to push one single object once. That flow, today, is impossible for that update policy.

Reply:
Fair. So I'd rather not implement this then. ( Like in the connect format. At least then it will be similar to the existing one. )